operator/v1/types_ingress.go

```diff
@@ -332,6 +340,10 @@ type IngressControllerStatus struct {
 	//     * DNS records have been successfully created.
 	//   - False if any of those conditions are unsatisfied.
 	Conditions []OperatorCondition `json:"conditions,omitempty"`
+
+	// tlsSecurityProfile is the observed TLS connection configuration.
+	// +optional
+	TLSSecurityProfile *configv1.TLSSecurityProfile `json:"tlsSecurityProfile,omitempty"`
```

Comment:
Shouldn't this be `*configv1.TLSProfileSpec`?

Reply:
Hm. I think so.

Reply:
Fixed

Reply:
Doesn't this mean we also need to add `TLSSecurityProfile` to `ingresses.config.openshift.io` like https://github.com/openshift/api/blob/master/config/v1/types_apiserver.go#L49 so that we can say our default is based on the cluster singleton?

Reply:
I don't see how that follows.  However, it would make sense to have some cluster-wide default TLS configuration.  The same default could be used for ingress, the API, and any other component where it would make sense.

Reply:
Added config to ingress.config.openshift.io.

Reply:
It follows the design laid out in https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/tls-config.md and applied to APIServerSpec. Where else could the defaults possibly come from?

Reply:
> It follows the design laid out in https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/tls-config.md and applied to APIServerSpec.

I agree that a requirement to have a cluster-wide setting follows from https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/tls-config.md (specifically, the statement, "This should be a shared, top level configuration amongst the affected components with a local override for ingress controllers").  I don't see how it follows from using `*configv1.TLSProfileSpec` for the value of this field.  However, in this case, it doesn't matter if we use different paths to arrive at the same conclusion.

> Where else could the defaults possibly come from?

Hard-coded defaults.  But using a cluster singleton is more obvious, flexible, and maintainable.

Reply:
Thinking about this some more, another option would be reading effective values from the [static mappings](https://github.com/openshift/api/blob/master/config/v1/types_tlssecurityprofile.go#L274) and never persisting them. In that case status would in fact contain a `TLSSecurityProfile` to avoid persisting the statically mapped spec data. I would like to see how @deads2k is planning on doing this for the APIServer.

Reply:
To further elaborate on what I'm saying:

Input:

```yaml
apiVersion: config.openshift.io/v1
kind: Ingress
spec:
  tlsSecurityProfile:
    type: Intermediate

apiVersion: operator.openshift.io/v1
kind: IngressController
spec:
  tlsSecurityProfile: nil
```

Output:

```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
spec:
  tlsSecurityProfile: nil
status:
  tlsProfileSpec:
    # from https://github.com/openshift/api/blob/master/config/v1/types_tlssecurityprofile.go#L313
    ciphers: # ...
    tlsVersion: # ...
    dhParamSize: # ...
```

Now, just looking at the ingresscontroller, what profile is it using? Assume the cluster config changed, so the original cluster default that led to the status is gone.

Instead:

```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
spec:
  tlsSecurityProfile: nil
status:
  tlsSecurityProfile:
    type: Intermediate
```

The ingress-operator uses the static mappings in `types_tlssecurityprofile.go` to convert this into a TLSSecuritySpec at runtime.

I suspect there are reasons for and against both ways I'm not considering so I wanted to clarify for discussion.

Reply:
I dislike the latter option because (1) the TLS profile of an existing ingresscontroller is changed _implicitly_ when the "Intermediate" profile changes, (2) the status is a lie during the period of time between when the definition of "Intermediate" changes and when the ingress operator reconciles the ingresscontroller, and (3) the status no longer says _explicitly_ what ciphers and TLS versions are enabled.

Reply:
>I dislike the latter option because (1) the TLS profile of an existing ingresscontroller is changed implicitly when the "Intermediate" profile changes, (2) the status is a lie during the period of time between when the definition of "Intermediate" changes and when the ingress operator reconciles the ingresscontroller, and (3) the status no longer says explicitly what ciphers and TLS versions are enabled.

Do you agree that the former way causes information loss? (e.g. the user's explicit intent to use the "Intermediate" profile.) If so, does it matter?

If the ingresscontroller is changed from Intermediate to Modern and the client misses the update and clears its cache, it won't have any knowledge of the Intermediate state (as it can't be inferred from status), but we should be level-based anyway.

I'm trying to think of a useful way to document the semantics... like, the profile type you specify is used to stamp down a generic point-in-time representation of the specific profile, and if you change the profile type on the existing ingresscontroller, it's possible a different config is stamped down even if the type is the same (i.e. we updated what "Intermediate" is).

Reply:
>I'm trying to think of a useful way to document the semantics... like, the profile type you specify is used to stamp down a generic point-in-time representation of the specific profile, and if you change the profile type on the existing ingresscontroller, it's possible a different config is stamped down even if the type is the same (i.e. we updated what "Intermediate" is).

My example here was wrong.. imagine if you specify Intermediate and then upgrade to a release where we change what Intermediate consists of. Do we rebuild your status and roll out automatically? If not, when, and how would we prevent that from happening the next time we reconcile for any other reason.

(Related to the mutability question?)

Reply:
> Do you agree that the former way causes information loss? (e.g. the user's explicit intent to use the "Intermediate" profile.) If so, does it matter?

The intent is in `spec`, so what is lost?

> If the ingresscontroller is changed from Intermediate to Modern and the client misses the update and clears its cache, it won't have any knowledge of the Intermediate state (as it can't be inferred from status), but we should be level-based anyway.

Which client are you talking about?  The operator's client, the client with which the administrator set `spec.tlsSecurityProfile`, or something else?

> My example here was wrong.. imagine if you specify Intermediate and then upgrade to a release where we change what Intermediate consists of. Do we rebuild your status and roll out automatically? If not, when, and how would we prevent that from happening the next time we reconcile for any other reason.

If `spec.tlsSecurityProfile` is mutable, I think the operator will have to updating the effective profile in this scenario.  Anything else would be an awful mess.  So let's tell the user to specify an explicit custom profile if ey needs it and otherwise accept that the profile ey specifies is subject to change.

Reply:
>The intent is in spec, so what is lost?

The prior intent.

>Which client are you talking about? The operator's client, the client with which the administrator set spec.tlsSecurityProfile, or something else?

Operator. But as I said (and taken with what you say next), I don't think it matters because we're talking about a level driven approach — the old intent doesn't matter. So never mind all that.

>If spec.tlsSecurityProfile is mutable, I think the operator will have to update the effective profile in this scenario. Anything else would be an awful mess. So let's tell the user to specify an explicit custom profile if ey needs it and otherwise accept that the profile ey specifies is subject to change.

👍 Okay, I think we're on the same page.

Reply:
Updated docs